distribution/pull_v2.go

```diff
@@ -501,6 +519,29 @@ func (p *v2Puller) pullSchema2(ctx context.Context, ref reference.Named, mfst *s
 		return imageID, manifestDigest, nil
 	}
 
+	var descriptors []xfer.DownloadDescriptor
+
+	// Note that the order of this loop is in the direction of bottom-most
+	// to top-most, so that the downloads slice gets ordered correctly.
+	for _, d := range mfst.Layers {
+		layerDescriptor := &v2LayerDescriptor{
+			digest:            d.Digest,
+			repo:              p.repo,
+			repoInfo:          p.repoInfo,
+			V2MetadataService: p.V2MetadataService,
+		}
+
+		if d.MediaType == schema2.MediaTypeForeignLayer && len(d.URLs) > 0 {
+			if runtime.GOOS != "windows" {
```

Comment:
> I think this is awesome for anyone willing to host their base layers somewhere else than the docker hub or any private registry.

I agree but that is no reason to hold up this PR. My point is that we need to be careful to not leak the foreign source abstraction, since it will make it more challenging to bring wider support here in the future. In other words, having foreign source here will make it harder to support this widely.